src/solution.c

```diff
@@ -285,40 +295,35 @@ void send_observations(u8 n, gps_time_t *t, navigation_measurement_t *m)
   }
 }
 
-static BinarySemaphore solution_wakeup_sem;
-#define tim5_isr Vector108
-#define NVIC_TIM5_IRQ 50
-void tim5_isr()
+static BSEMAPHORE_DECL(solution_wakeup_sem, TRUE);
+static void sol_gpt_cb(GPTDriver *gptd)
 {
-  CH_IRQ_PROLOGUE();
-  chSysLockFromIsr();
+  (void)gptd;
+  chSysLockFromISR();
 
   /* Wake up processing thread */
   chBSemSignalI(&solution_wakeup_sem);
 
-  timer_clear_flag(TIM5, TIM_SR_UIF);
-
-  chSysUnlockFromIsr();
-  CH_IRQ_EPILOGUE();
+  chSysUnlockFromISR();
 }
 
-static void timer_set_period_check(uint32_t timer_peripheral, uint32_t period)
+static void timer_set_period_check(GPTDriver *gptd, uint32_t period)
 {
-  __asm__("CPSID i;");
-  TIM_ARR(timer_peripheral) = period;
-  uint32_t tmp = TIM_CNT(timer_peripheral);
+  chSysLock();
+  gptChangeIntervalI(gptd, period);
+  u32 tmp = gptGetCounterX(gptd);
   if (tmp > period) {
-    TIM_CNT(timer_peripheral) = period;
+    gpt_lld_set_counter(gptd, period);
```

Comment:
What happens if the interval is set to a value less than the current counter?  This check is meant to catch that condition.  This code is not portable anyway, the meaning of the counter value is explicitly undefined.